developer/services/disposablevm-implementation.rst

```diff
@@ -2,73 +2,162 @@
 
-   2. (in APPVM) the amount of used memory is stored to ``qubes_used_mem`` xenstore key
+The user desires to circumvent any slow process, the creation of disposables fits into this category. Preloaded disposables enables fast retrieval of fresh disposables, so users don't have to get away from the computer or switch tasks when requesting disposables (or not requesting one at all because it was slow).
 
-   3. (in APPVM) busy-waiting for ``qubes_restore_complete`` xenstore key to appear
+Preloaded disposable stages
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
+There are several stages a disposable goes through while preloading and being used. In short:
 
+1. **Preload**: The qube is created and marked as preloaded. Qube is not visible in GUI applications.
 
-6. when ``qubes_used_mem`` key appears, the domain memory is reduced to this amount, to make the savefile smaller.
+  #. **Startup**: Begins qube startup, start basic services in it and attempt to interrupt (suspend/pause).
 
-7. APPVM private image is detached
+  #. **Request**: The qube is removed from the preload list. If startup has not yet reached interrupt, the latter is skipped.
 
-8. the domain is saved via ``xl save``
+2. **Used**: The qube is marked as used and may be unpaused/resumed (if applicable). Only in this phase, GUI applications treat the qube as any other unnamed disposable and the qube object is returned to the caller if requested.
 
-9. the COW file volatile.img (cow for root fs and swap) is packed to ``saved_cows.tar`` archive
+Preloaded disposable's worry-free life-cycle
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 
+There are several events that may trigger the creation or deletion of preloaded disposables, these events are dealt with methods decorated with :py:func:`qubes.events.handler <core-admin:qubes.events.handler>`. These events happen for different reasons, mostly because:
 
-The ``qubes_prepare_saved_domain.sh`` script is lowlevel. It is usually called by ``qvm-create-default-dvm`` script, that takes care of creating a special AppVM (named template_name-dvm) to be passed to ``qubes_prepare_saved_domain.sh``, as well as copying the savefile to /dev/shm (the latter action is not done if the ``/var/lib/qubes/dvmdata/dont_use_shm`` file exists).
+- There is a gap between the current number of preloaded disposables and the maximum number of preloaded disposables allowed; or
+- The current number of preloaded disposable is over the maximum allowed.
 
-Restoring a DisposableVM from the savefile
-------------------------------------------
+The preload creation can also fail for different reasons:
 
+- Qubesd was interrupted mid preload creation, on the next service restart, :py:meth:`domain-load <core-admin:qubes.vm.mix.dvmtemplate.DVMTemplateMixin.on_domain_loaded>` of the disposable template will refresh the incomplete disposables;
+- Service to check if the system is fully operation has failed; and
+- There is not enough memory to preload at the moment.
```

Comment:
This part looks out of order - before you talk when preloading happens, after you talk about the same, but here you have some common failure modes.